Add Stable Diffusion example (#178)
opendatahub-io/distributed-workloads: examples/stable-diffusion-dreambooth/README.md

## Introduction

This example demostrate how to fine tune Stable Diffusion with DreamBooth and Ray Train using OpenShift AI.  

## Credits

The example codes are from:
1. [Ray dreambooth finetuning](https://docs.ray.io/en/latest/train/examples/pytorch/dreambooth_finetuning.html)
1. [Ray Train Stable Diffussion Notebook](https://github.com/GoogleCloudPlatform/ai-on-gke/blob/main/ray-on-gke/examples/notebooks/raytrain-stablediffusion.ipynb)

## Overview 

> In this example, a workbench will be automatically created in this example with a few notebooks to explore Ray.

Reviewer: ```suggestion
In this example, a workbench will be automatically created with a few notebooks to explore Ray.
```

| Name | Description |
| -------------------- | ----------- |

> | [Introduction](notebook/00%20Intro.ipynb) | A simple demostrate of ray remote function         |

Reviewer: ```suggestion
| [Introduction](notebook/00%20Intro.ipynb) | A simple demonstration of the ray remote function         |
```

> | [Ray Train with Stable Diffusion](notebook/01%20raytrain-stablediffusion.ipynb) | Using Ray train to distibute training

Reviewer: ```suggestion
| [Ray Train with Stable Diffusion](notebook/01%20raytrain-stablediffusion.ipynb) | Using Ray Train to distribute training
```

| [Archiving and uploading model](notebook/02%20raytrain-stablediffusion-upload-model.ipynb) | Using Torch model archiver and upload to S3 bucket
| [Serving using KServe](notebook/03%20raytrain-stablediffusion-model-serving.ipynb) | Setting up KServe with PyTorch |
| [Text-to-Image Chatbot](notebook/04%20raytrain-stablediffusion-streamlit.ipynb) | Deploying Streamlit chat app 

## Prerequisite

> * Ensure there is at least 1 worker node that has a GPU. On AWS, this can be a p3.8xlarge instance, otherwise you can run the makefile target to add a `machineset` for a single replica of p3.8xlarge.

Reviewer: the p3.8xlarge has 4 gpus. Are they all required?

  ```bash
  make add-gpu-machineset
  ```

  Taint the GPU node
  ```bash
  oc adm taint nodes <gpu-node> nvidia.com/gpu=Exists:NoSchedule
  ```

* Ensure there is a RWX available storage class, such as `
ocs-storagecluster-cephfs`. The `stable-diffusion-shared-storage` pvc uses cephfs otherwise update the PVC resource. 

  You can install OpenShift Data Foundation (ODF) to achieve this. ODF will require a minimum of 3 worker nodes to start with. 

  ```
  make deploy-odf
  ```

  The CSI plugin daemonset will have to be updated to tolerate the GPU taint. 

  ``` 
  oc apply -f yaml/operators/rook-ceph-operator-config.yaml

>   ```

Reviewer: I ran through the setup a few times and I only have the following DataClasses available:
![image](https://github.com/user-attachments/assets/fc6190ed-62f4-4a2e-affc-ec864d9d70c5)
causing the `git-clone-job` to hang waiting for a PVC

  ``` yaml
  data:
    CSI_PLUGIN_TOLERATIONS: |-
      - key: nvidia.com/gpu
        operator: Exists
        effect: NoSchedule
  ```

## Setup
Install OpenShift AI using the OpenShift AI Operator. This install the latest version from the fast channel.

```bash
make install-openshift-ai
```

## Setting Up the Demo

Run the makefile target. This creates a Data Science Project called `distributed` with the following
* A CephFS RWX PVC
* A single workbench with the git repository already cloned inside
* Kueue configurations

``` bash

> make setup-ray-distributed-training

Reviewer: Can you also make a warning that this will delete all kueue resources and RayClusters. Is deleting all of them necessary for the example?
